crates/sdk/src/masp/utils.rs

```diff
@@ -153,13 +245,426 @@ where
+                             {from_height}-{to_height}: {err}"
+                        ))
+                    })?;
+                if !response.status().is_success() {
+                    let err = Self::get_server_error(response).await?;
+                    return Err(Error::Other(format!(
+                        "Failed to fetch transactions in the range \
+                         {from_height}-{to_height}: {err}"
+                    )));
+                }
+                let payload: TxResponse =
+                    response.json().await.map_err(|err| {
+                        Error::Other(format!(
+                            "Could not deserialize the transactions JSON \
+                             response in the height range \
+                             {from_height}-{to_height}: {err}"
+                        ))
+                    })?;
+
+                Ok::<_, Error>(payload)
+            };
+            let payload = tokio::select! {
+                maybe_payload = txs_fut => {
+                    maybe_payload?
+                }
+                _ = &mut *shutdown_signal => {
+                    return Err(Error::Interrupt(
+                        "[ShieldedSync::Fetching]".to_string(),
+                    ));
+                }
+            };
+
+            let mut last_height = None;
+
+            for Transaction {
+                batch,
+                block_index,
+                block_height,
+            } in payload.txs
+            {
+                let mut extracted_masp_txs = Vec::with_capacity(batch.len());
+
+                for TransactionSlot { bytes } in batch {
+                    type MaspTx = masp_primitives::transaction::Transaction;
+
+                    extracted_masp_txs.push(
+                        MaspTx::try_from_slice(&bytes).map_err(|err| {
+                            Error::Other(format!(
+                                "Could not deserialize the masp txs borsh \
+                                 data at height {block_height} and index \
+                                 {block_index}: {err}"
+                            ))
+                        })?,
+                    );
+                }
+
+                tx_sender.send((
+                    IndexedTx {
+                        height: BlockHeight(block_height),
+                        index: TxIndex(block_index),
+                    },
+                    extracted_masp_txs,
+                ));
+
+                let curr_height = Some(block_height);
+                if curr_height > last_height {
+                    last_height = curr_height;
+                    _ = fetch_iter.next();
+                }
+            }
+
+            if from >= to {
+                break;
+            }
+        }
+
+        // NB: a hack to drain the iterator in case
+        // there were still some items left
+        // TODO: improve this lol
+        while fetch_iter.next().is_some() {}
```

Comment:
Couldn't this be a sign that something went wrong?

Reply:
this iterator is going over a range of blocks. if we do not find txs in some specific block, the iterator will not be moved forward, resulting in some elements left to yield after we exit the first loop. this is just a hack to drain the iterator in these cases.

we will get rid of this mechanic in the phase 3 code, progress will be kept track of differently